### PR TITLE
Feature/api password recovery

### PR DESCRIPTION
**Pull request by btry**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | not added yet
| Fixed tickets | #1314

This PR adds and endpoint to recover an account.

By the way, fixes some links in the rest API documentation

### CODE REVIEW

File: apirest.md [hunk 1]
```diff
@@ -168,6 +169,26 @@ $ curl -X GET \
...
+* **Parameters**: (query string)
+  * *email*: email address of the user to recover. Mandatory.
+* **Returns**:
+  * 200 (OK).
+  * 400 (Bad Request) with a message indicating an error in input parameter.
+
+```bash
+$ curl -X GET \
+-H 'Content-Type: application/json' \
+'http://path/to/glpi/apirest.php/lostPassword'
```
Comment (orthagh) [inline apirest.md]: we miss the `email` parameter in the example

File: inc/api.class.php [hunk 2]
```diff
@@ -1817,6 +1817,77 @@ protected function deleteItems($itemtype, $params = []) {
...
+
+      if (!isset($params['email'])) {
+         return $this->returnError(__("email parameter missing"));
+      }
+
+      if (isset($_SESSION['glpiID'])) {
+         return $this->returnError(__("A session is active"));
+      }
+
+      $user = new User();
```
Comment (orthagh) [inline inc/api.class.php]: I think this part need to be factorised with User::updateForgottenPassword

File: inc/apirest.class.php [hunk 3]
```diff
@@ -189,6 +189,13 @@ public function call() {
 
          return $this->returnResponse($response, $code, $additionalheaders);
 
+      } else if ($resource === "lostPassword") {
+         if ($this->verb != 'POST' && $this->verb != 'PUT') {
+            // forbid password reset with GET because it may be logged
+            return $this->returnError();
```
Comment (orthagh) [inline inc/apirest.class.php]: maybe precise the error cause

Comment (btry) [inline inc/apirest.class.php]: ReturnError uses 'Bad Request' by default. Do you think this is not sufficient?

Comment (orthagh) [inline inc/apirest.class.php]: maybe precise that this endpoint couldn't be used by GET.
By the way, you have also an inconsistence with the inline documentation which precise GET verb for this endpoint

Comment (btry) [inline inc/apirest.class.php]: OK to send a specific message.

I initially thought GET would be suffficient. I have to update the documentation.

File: inc/apirest.class.php [hunk 4]
```diff
@@ -189,6 +189,13 @@ public function call() {
 
          return $this->returnResponse($response, $code, $additionalheaders);
 
+      } else if ($resource === "lostPassword") {
+         if ($this->verb != 'PUT') {
+            // forbid password reset with GET because it may be logged
```
Comment (orthagh) [inline inc/apirest.class.php]: comment inconsistent with the relative code

File: inc/config.class.php [hunk 5]
```diff
@@ -38,6 +38,20 @@
    die("Sorry. You can't access this file directly");
 }
 
+
+class ConfigPasswordTooWeak extends Exception {
```
Comment (orthagh) [inline inc/config.class.php]: I'm not sure about the Exception class in config file.
Maybe we could add a new sub exceptions directory into inc to store this first one (and next).

File: inc/user.class.php [hunk 6]
```diff
@@ -41,6 +41,10 @@
 
 use Sabre\VObject;
 
+class UserForgetPasswordException extends Exception {
```
Comment (orthagh) [inline inc/user.class.php]: same for exception

File: inc/config.class.php (hunk 5 above)
Comment (btry) [inline inc/config.class.php]: I agree.

File: tests/APIBaseClass.php [hunk 7]
```diff
@@ -1280,6 +1280,98 @@ public function testKillSession() {
    }
 
    /**
+    * @tags api
+    * @covers API::lostPassword
```
Comment (trasher) [inline tests/APIBaseClass.php]: There is no `@covers` with atoum ;)

File: tests/APIBaseClass.php [hunk 8]
```diff
@@ -1280,6 +1280,98 @@ public function testKillSession() {
...
+                          'ERROR');
+
+      // Test the verb GET is not alloxed
+      $res = $this->query('lostPassword',
+                          ['verb'    => 'GET',
+                          ],
+                          400,
+                          'ERROR');
+
+      // Test the verb DELETE is not alloxed
```
Comment (trasher) [inline tests/APIBaseClass.php]: Typo

**Comment by trasher** [comment]
Looks globally OK but updatePassword methods needs to be tested as well (success and exceptions).
Also, I'm not sure exceptions are tested from API.

**Comment by btry** [comment]
I had to add in  the API an new handler for thie endpoint, and it catches the exceptions. See api.class.php line 1860.

@covers and typo fixes incoming

**Pull request merged**